### DISCUSSION TITLE
[TinyUSB] USB in host mode

### DISCUSSION THREAD

**Discussion by ayoubSoussi** [modm-io/modm#903] [Q&A] [answered]
Hello everyone,

I'm trying to use modm library on my STM32F4 discovery board that has a USB OTG, but I need to use TinyUSB in host mode to be able to connect (and make use of) a USB flash drive to my board.

After reading the TinyUSB section of the modm documentation, I think only device mode is currently implemented. Can you please confirm that for me and if you have any hints on how to implement host mode with modm ? I appreciate your help !

**Comment by salkinium** [comment]
Device mode is especially supported, but host most should still work if you leave the `modm:tinyusb:config` option empty and instead manually include the host modules (eg. `modm:tinyusb:host:cdc`) and configure them using the `<tusb_config_local.h>` file.
You may also need to provide a USB descriptor manually.

I would just try it and compare what is missing from the TinyUSB host examples. Note that TinyUSB v0.14 was just released this week, but modm is still using v0.13 which is quite old now.

**Reply by salkinium** [reply]
FYI: I haven't tested host mode yet, so there may be some further pitfalls, I strongly suggest you compile with `profile=debug` and the `CFG_TUSB_DEBUG=2` as described in the docs. I also think you need to redirect the `modm_abandon` implementation to use a serial logger or RTT.

**Reply by salkinium** [reply]
Also verify that device mode works first on the F4 discovery, so that you can exclude issues with the hardware connection setup, especially the Vusb signal which can be tricky to setup correctly.

**Reply by ayoubSoussi** [reply]
I checked device mode on my F4 discovery board, it works !

**Comment by ayoubSoussi** [comment, marked as answer]
I had a quick look at the TinyUSB documentation, and apparently TinyUSB doesn't support yet Host mode for the F4 family of STM32 MCUs.

**Reply by salkinium** [reply]
Oh, that's a bummer. That was probably the reason why there are no host examples…

**Comment by ayoubSoussi** [comment]
So as an alternative, I wanna use an SD card in SPI mode to store data. Is there a driver for that in modm ? (I looked and apparently there isn't) or do I need to develop it myself on top of the SPI module ?

**Reply by salkinium** [reply]
No, we don't have a driver yet, it's been on my todo list forever, just never got to it.
